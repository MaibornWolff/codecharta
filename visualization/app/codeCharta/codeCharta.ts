"use strict";

import "./core/core.module";
import "./ui/ui";

import {codeChartaComponent, CodeChartaController} from "./codeCharta.component";

import angular from "angular";

angular.module(
    "app.codeCharta",
    ["app.codeCharta.core", "app.codeCharta.ui"]
);

angular.module("app.codeCharta").component(
    codeChartaComponent.selector,
    codeChartaComponent
);


