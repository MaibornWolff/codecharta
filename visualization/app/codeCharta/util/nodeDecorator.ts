"use strict"
import * as d3 from "d3"
import { HierarchyNode } from "d3"
import {
	BlacklistItem,
	BlacklistType,
	CCFile,
	CodeMapNode,
	MetricData,
	FileMeta,
	EdgeMetricCount,
	KeyValuePair,
	AttributeTypes,
	AttributeTypeValue
} from "../codeCharta.model"
import { CodeMapHelper } from "./codeMapHelper"
import { MetricService } from "../state/metric.service"

export class NodeDecorator {
	public static decorateMap(map: CodeMapNode, fileMeta: FileMeta, metricData: MetricData[]) {
		this.decorateMapWithMissingObjects(map, fileMeta)
		this.decorateMapWithCompactMiddlePackages(map)
		this.decorateLeavesWithMissingMetrics(map, metricData)
	}

	public static preDecorateFile(file: CCFile) {
		this.decorateMapWithPathAttribute(file)
	}

	private static decorateMapWithCompactMiddlePackages(map: CodeMapNode) {
		const isEmptyMiddlePackage = current => {
			return (
				current &&
				current.children &&
				current.children.length === 1 &&
				current.children[0].children &&
				current.children[0].children.length > 0
			)
		}

		const rec = current => {
			if (isEmptyMiddlePackage(current)) {
				let child = current.children[0]
				current.children = child.children
				current.name += "/" + child.name
				current.path += "/" + child.name
				if (child.link) {
					current.link = child.link
				}
				current.attributes = child.attributes
				current.edgeAttributes = child.edgeAttributes
				current.deltas = child.deltas
				rec(current)
			} else if (current && current.children && current.children.length > 1) {
				for (let i = 0; i < current.children.length; i++) {
					rec(current.children[i])
				}
			}
		}

		if (map) {
			rec(map)
		}
	}

	private static decorateMapWithPathAttribute(file: CCFile) {
		if (file && file.map) {
			let root = d3.hierarchy<CodeMapNode>(file.map)
			root.each(node => {
				node.data.path =
					"/" +
					root
						.path(node)
						.map(x => x.data.name)
						.join("/")
			})
		}
		return file
	}

	private static decorateMapWithMissingObjects(map: CodeMapNode, fileMeta: FileMeta) {
		if (map) {
			let root = d3.hierarchy<CodeMapNode>(map)
			root.each(node => {
				node.data.visible = true
				node.data.attributes = !node.data.attributes ? {} : node.data.attributes
				node.data.edgeAttributes = !node.data.edgeAttributes ? {} : node.data.edgeAttributes
				Object.assign(node.data.attributes, { [MetricService.UNARY_METRIC]: 1 })
			})
		}
	}

	private static decorateLeavesWithMissingMetrics(map: CodeMapNode, metricData: MetricData[]) {
		if (map && metricData) {
			let root = d3.hierarchy<CodeMapNode>(map)
			root.leaves().forEach(node => {
				metricData.forEach(metric => {
					if (!node.data.attributes.hasOwnProperty(metric.name)) {
						node.data.attributes[metric.name] = 0
					}
				})
			})
		}
	}

	public static decorateParentNodesWithAggregatedAttributes(
		map: CodeMapNode,
		blacklist: BlacklistItem[],
		metricData: MetricData[],
		edgeMetricData: MetricData[],
		isDeltaState: boolean,
		attributeTypes: AttributeTypes
	) {
		if (map) {
			let root = d3.hierarchy<CodeMapNode>(map)
			root.each((node: HierarchyNode<CodeMapNode>) => {
				const leaves: HierarchyNode<CodeMapNode>[] = node
					.leaves()
					.filter(x => !CodeMapHelper.isBlacklisted(x.data, blacklist, BlacklistType.exclude))
				this.decorateNodeWithAggregatedChildrenMetrics(leaves, node, metricData, isDeltaState, attributeTypes)
				this.decorateNodeWithChildrenSumEdgeMetrics(leaves, node, edgeMetricData, attributeTypes)
			})
		}
		return map
	}

	private static decorateNodeWithAggregatedChildrenMetrics(
		leaves: HierarchyNode<CodeMapNode>[],
		node: HierarchyNode<CodeMapNode>,
		metricData: MetricData[],
		isDeltaState: boolean,
		attributeTypes: AttributeTypes
	) {
		metricData.forEach(metric => {
			if (node.data.children && node.data.children.length > 0) {
				node.data.attributes[metric.name] = this.aggregateLeafMetric(
					leaves.map(x => x.data.attributes),
					metric.name,
					attributeTypes
				)
				if (isDeltaState) {
					node.data.deltas[metric.name] = this.aggregateLeafMetric(leaves.map(x => x.data.deltas), metric.name, attributeTypes)
				}
			}
		})
	}

	private static decorateNodeWithChildrenSumEdgeMetrics(
		leaves: HierarchyNode<CodeMapNode>[],
		node: HierarchyNode<CodeMapNode>,
		edgeMetricData: MetricData[],
		attributeTypes: AttributeTypes
	) {
		edgeMetricData.forEach(edgeMetric => {
			if (node.data.children && node.data.children.length > 0) {
				node.data.edgeAttributes[edgeMetric.name] = this.aggregateLeafEdgeMetric(leaves, edgeMetric.name, attributeTypes)
			}
		})
	}

	private static aggregateLeafMetric(metrics: KeyValuePair[], metricName: string, attributeTypes: AttributeTypes): number {
		const metricValues: number[] = metrics.map(x => x[metricName]).filter(x => !!x)
		const attributeType = attributeTypes.nodes[metricName]

		if (metricValues.length == 0) {
			return 0
		}

		switch (attributeType) {
			case AttributeTypeValue.relative:
				return this.median(metricValues)
			case AttributeTypeValue.absolute:
			default:
				return metricValues.reduce((partialSum, a) => partialSum + a)
		}
	}

	private static aggregateLeafEdgeMetric(
		leaves: HierarchyNode<CodeMapNode>[],
		metricName: string,
		attributeTypes: AttributeTypes
	): EdgeMetricCount {
		const metricValues: EdgeMetricCount[] = leaves.map(x => x.data.edgeAttributes[metricName]).filter(x => !!x)
		const attributeType = attributeTypes.edges[metricName]

		const concentrated = { incoming: [], outgoing: [] }
		metricValues.forEach(element => {
			concentrated.incoming.push(element.incoming)
			concentrated.outgoing.push(element.outgoing)
		})

		if (metricValues.length == 0) {
			return { incoming: 0, outgoing: 0 }
		}

		switch (attributeType) {
			case AttributeTypeValue.relative:
				return { incoming: this.median(concentrated.incoming), outgoing: this.median(concentrated.outgoing) }
			case AttributeTypeValue.absolute:
			default:
				return {
					incoming: concentrated.incoming.reduce((pS, a) => pS + a),
					outgoing: concentrated.outgoing.reduce((pS, a) => pS + a)
				}
		}
	}

	private static median(numbers: number[]): number {
		const middle = (numbers.length - 1) / 2
		numbers.sort((a, b) => a - b)
		return (numbers[Math.floor(middle)] + numbers[Math.ceil(middle)]) / 2
	}
}
