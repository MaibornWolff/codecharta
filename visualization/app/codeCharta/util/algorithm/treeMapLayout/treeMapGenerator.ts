import { hierarchy, HierarchyNode, HierarchyRectangularNode, treemap } from "d3-hierarchy"
import { TreeMapHelper, treeMapSize } from "./treeMapHelper"
import { CodeMapNode, DynamicSettings, Node, NodeMetricData, State } from "../../../codeCharta.model"
import { getMapResolutionScaleFactor, isLeaf } from "../../codeMapHelper"
import { calculatePaddingBasedOnBuildingArea } from "./paddingCalculator"
import { getChildrenAreaValues, getSmallestValueOrSmallestDifference } from "./modifiedTreeMapHelperFunctions"
import { calculateTotalNodeArea } from "./nodeAreaCalculator"

export type SquarifiedTreeMap = { treeMap: HierarchyRectangularNode<CodeMapNode>; height: number; width: number }

export const FOLDER_LABEL_TOO_SMALL_PARENT = 0.2
export const FOLDER_LABEL_TOO_SMALL_ROUTE = 0.02
export const INITIAL_PADDING = 50
export const DEFAULT_PADDING_FLOOR_LABEL_FROM_LEVEL_0 = 100
export const DEFAULT_PADDING_FLOOR_LABEL_FROM_LEVEL_1 = 85
export const PADDING_APPROX_FOR_DEPTH_ZERO = 0.035
export const PADDING_APPROX_FOR_DEPTH_ONE = 0.028

const MIN_BUILDING_AREA = 100
const PADDING_SCALING_FACTOR = 0.4

export function createTreemapNodes(map: CodeMapNode, state: State, metricData: NodeMetricData[], isDeltaState: boolean): Node[] {
	const mapSizeResolutionScaling = getMapResolutionScaleFactor(state.files)
	const maxHeight = metricData.find(x => x.name === state.dynamicSettings.heightMetric).maxValue * mapSizeResolutionScaling
	const maxWidth = metricData.find(x => x.name === state.dynamicSettings.areaMetric).maxValue * mapSizeResolutionScaling
	const heightScale = (treeMapSize * 2) / maxHeight
	if (hasFixedFolders(map)) {
		const scaledPadding = (INITIAL_PADDING * state.dynamicSettings.margin) / 100
		const hierarchyNode = hierarchy(map)
		// Base root folder has width: 100px and length: 100px
		const nodes: Node[] = [TreeMapHelper.buildRootFolderForFixedFolders(hierarchyNode.data, heightScale, state, isDeltaState)] // nosonar
		// Multiply mapSize of (default) 250px by 2 = 500px and add the total margin
		const totalMapSize = treeMapSize * 2 + getEstimatedNodesPerSide(hierarchyNode) * (scaledPadding / PADDING_SCALING_FACTOR)
		// than divide through the root folder width and length to get a scale factor for calculation for all following nodes.
		const scaleLength = totalMapSize / nodes[0].width
		const scaleWidth = totalMapSize / nodes[0].length
		// Scale the 100x100 root folder to be bigger and to match fixed/estimated totalMapSize
		scaleRoot(nodes[0], scaleLength, scaleWidth)
		return [
			...nodes,
			...buildSquarifiedTreeMapsForFixedFolders(
				hierarchyNode,
				state,
				scaleLength,
				scaleWidth,
				0,
				0,
				heightScale,
				maxHeight,
				maxWidth,
				isDeltaState,
				mapSizeResolutionScaling
			)
		]
	}

	const squarifiedTreeMap = getSquarifiedTreeMap(map, state)

	const nodes: Node[] = []
	for (const squarifiedNode of squarifiedTreeMap.treeMap) {
		nodes.push(TreeMapHelper.buildNodeFrom(squarifiedNode, heightScale, maxHeight, state, isDeltaState))
	}
	return nodes
}
function buildSquarifiedTreeMapsForFixedFolders(
	hierarchyNode: HierarchyNode<CodeMapNode>,
	state: State,
	scaleLength: number,
	scaleWidth: number,
	offsetX0: number,
	offsetY0: number,
	heightScale: number,
	maxHeight: number,
	maxWidth: number,
	isDeltaState: boolean,
	mapSizeResolutionScaling: number
) {
	const nodes = []

	for (const fixedFolder of hierarchyNode.children) {
		const fixedPosition = fixedFolder.data.fixedPosition
		const squarified = getSquarifiedTreeMap(fixedFolder.data, state)

		for (const squarifiedNode of squarified.treeMap.descendants()) {
			// squarified.width/height is a sum of the fixedPosition.width/height and applied margins
			// The following scaling factors are used later to calculate out the original percentage fixed width/height.
			// Example: Width and Height fixed at 20. Added margins 42px (absolute value) for each = 62px
			//  20/62 = 0.322...       62 x 0.322... = 20 :)
			const scaleX = fixedPosition.width / squarified.width
			const scaleY = fixedPosition.height / squarified.height
			// Scale the fixed (percentage) positions of a node by the right scale factor, so that it will be placed properly.
			// The squarifiedNode coordinates x0, x1, y0, y1 are already assigned with positions from the treemap algorithm.
			squarifiedNode.x0 = (squarifiedNode.x0 * scaleX + fixedPosition.left) * scaleWidth
			squarifiedNode.x1 = (squarifiedNode.x1 * scaleX + fixedPosition.left) * scaleWidth
			squarifiedNode.y0 = (squarifiedNode.y0 * scaleY + fixedPosition.top) * scaleLength
			squarifiedNode.y1 = (squarifiedNode.y1 * scaleY + fixedPosition.top) * scaleLength
			// Add x and y (absolute px) offsets from parent fixed folder, if any.
			squarifiedNode.x0 += offsetX0
			squarifiedNode.x1 += offsetX0
			squarifiedNode.y0 += offsetY0
			squarifiedNode.y1 += offsetY0
			const node = TreeMapHelper.buildNodeFrom(squarifiedNode, heightScale, maxHeight, state, isDeltaState, true)
			nodes.push(node)
			if (hasFixedFolders(fixedFolder.data)) {
				// Imagine the parent Folder has absolute px-width of 341px
				// We need to calculate a scale factor for the fixed child to transform a relative percentage fixed width/height to px.
				// Example: 20% of 341px:
				//  fixedWidth=20, parent Fixed Folder width: 341px => 20 * (scaleLength=341/100)
				// The original scaleWidth/scaleLength must be retained for the case that you have more than one fixedFolder on the same level.
				const childRelativeLengthScale = node.length / 100
				const childRelativeWidthScale = node.width / 100
				Array.prototype.push.apply(
					nodes,
					buildSquarifiedTreeMapsForFixedFolders(
						fixedFolder,
						state,
						childRelativeLengthScale,
						childRelativeWidthScale,
						squarifiedNode.x0,
						squarifiedNode.y0,
						heightScale,
						maxHeight,
						maxWidth,
						isDeltaState,
						mapSizeResolutionScaling
					)
				)
				// the break is actually needed!
				// the inner for-loop loops over the (parent) fixedFolder and its descendants.
				// if a direct child is a fixedFolder as well it will be handled by the recursive function call.
				// In this case, we must break the inner loop after handling the fixedFolder child recursively
				// and therefore prevent that the fixed child will be added as a node twice.
				break
			}
		}
	}
	return nodes
}
function hasFixedFolders(map: CodeMapNode) {
	// What if the second child of root is the first fixed folder?
	// Assumption: all children of the root folder require the fixedPosition attribute.
	return Boolean(map.children[0]?.fixedPosition)
}
function scaleRoot(root: Node, scaleLength: number, scaleWidth: number) {
	root.x0 *= scaleWidth
	root.y0 *= scaleLength
	root.width *= scaleWidth
	root.length *= scaleLength
}

export function calculateBuildingAreaWithProportionalPadding(area: number, smallestDelta: number, minimumBuildingArea: number) {
	if (area === 0) {
		return 0
	}

	return (area / smallestDelta) * minimumBuildingArea
}

export function getBuildingAreasWithProportionalPadding(childrenAreaValues: number[], smallestDelta: number, minimumBuildingSize: number) {
	return childrenAreaValues.map(element => {
		return calculateBuildingAreaWithProportionalPadding(element, smallestDelta, minimumBuildingSize)
	})
}

function getSquarifiedTreeMap(map: CodeMapNode, state: State): SquarifiedTreeMap {
	map.children = map.children.filter(child => child.attributes[state.dynamicSettings.areaMetric] !== 0)

	const hierarchyNode = hierarchy(map)
	let padding = INITIAL_PADDING * (state.dynamicSettings.margin / 100)

	// from hierarchy Node to areas
	const childrenAreaValues = getChildrenAreaValues(hierarchyNode, state)

	const smallestDelta = getSmallestValueOrSmallestDifference(childrenAreaValues)

	padding = calculatePaddingBasedOnBuildingArea(childrenAreaValues, smallestDelta, MIN_BUILDING_AREA, padding)

	const metricBuildingAreasIncludingPadding = getBuildingAreasWithProportionalPadding(
		childrenAreaValues,
		smallestDelta,
		MIN_BUILDING_AREA
	)

	const { rootWidth, rootHeight, metricSum } = calculateTotalNodeArea(metricBuildingAreasIncludingPadding, hierarchyNode, padding, state)

	const width = rootWidth
	const height = rootHeight
	let rootNode

	const treeMap = treemap<CodeMapNode>()
		.size([height, width]) // padding does not need to be divided by 2, already calculated by d3-hierarchy
		.paddingOuter(padding) // padding of the element; applied to leaves
		.paddingInner(padding) // margin of the element; applied to folders
		.paddingRight(node => {
			if (!rootNode && node.parent === null) {
				rootNode = node
			}
			// it seems that depth property is missing in that case
			// so the default padding will be added, which is fine though.
			if (rootNode) {
				const rootSize = rootNode.x1 - rootNode.x0

				// Start the labels at level 1 not 0 because the root folder should not be labeled
				if (node.depth === 0) {
					// Add a big padding for the first folder level (the font is bigger than in deeper levels)
					return Math.max(rootSize * PADDING_APPROX_FOR_DEPTH_ZERO, DEFAULT_PADDING_FLOOR_LABEL_FROM_LEVEL_0)
				}
				if (
					node.depth > 0 &&
					node.depth < 3 &&
					node.value / node.parent.value > FOLDER_LABEL_TOO_SMALL_PARENT &&
					node.value / rootNode.value > FOLDER_LABEL_TOO_SMALL_ROUTE
				) {
					return Math.max(rootSize * PADDING_APPROX_FOR_DEPTH_ONE, DEFAULT_PADDING_FLOOR_LABEL_FROM_LEVEL_1)
				}
			}

			// add treemap algorithm default padding otherwise
			return padding
		})

	const logSum = treeMap(metricSum)

	return { treeMap: logSum, height, width }
}

function getEstimatedNodesPerSide(hierarchyNode: HierarchyNode<CodeMapNode>) {
	let totalNodes = 0
	let blacklistedNodes = 0
	hierarchyNode.each(({ data }) => {
		if (data.isExcluded || data.isFlattened) {
			blacklistedNodes++
		}
		totalNodes++
	})
	// What does this line do?
	// Imagine a 3x3 grid of 9 nodes
	// 3 nodes are placed on the x-axis and 3 on the y-axis = 6
	// The calculated value is probably used to calculate the total margin which extends length and width of the map.
	return 2 * Math.sqrt(totalNodes - blacklistedNodes)
}
function isOnlyVisibleInComparisonMap(node: CodeMapNode, dynamicSettings: DynamicSettings) {
	return node.attributes[dynamicSettings.areaMetric] === 0 && node.deltas[dynamicSettings.heightMetric] < 0
}

// Only exported for testing.
export function getAreaValue(node: CodeMapNode, { dynamicSettings }: State) {
	if (node.isExcluded) {
		return 0
	}
	if (node.deltas && isOnlyVisibleInComparisonMap(node, dynamicSettings)) {
		return Math.abs(node.deltas[dynamicSettings.areaMetric])
	}

	if (isLeaf(node) && node.attributes?.[dynamicSettings.areaMetric]) {
		return node.attributes[dynamicSettings.areaMetric]
	}
	return 0
}
