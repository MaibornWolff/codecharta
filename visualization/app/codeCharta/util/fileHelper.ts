import { ExportBlacklistType, ExportCCFile, OldAttributeTypes } from "../codeCharta.api.model"
import { AttributeTypes, BlacklistItem, BlacklistType, CCFile } from "../codeCharta.model"

export function getCCFile(fileName: string, fileContent: ExportCCFile): CCFile {
	return {
		fileMeta: {
			fileName,
			fileChecksum: fileContent.fileChecksum,
			projectName: fileContent.projectName,
			apiVersion: fileContent.apiVersion
		},
		settings: {
			fileSettings: {
				edges: fileContent.edges || [],
				attributeTypes: getAttributeTypes(fileContent.attributeTypes),
				blacklist: potentiallyUpdateBlacklistTypes(fileContent.blacklist || []),
				markedPackages: fileContent.markedPackages || []
			}
		},
		map: fileContent.nodes[0]
	}
}

function getAttributeTypes(attributeTypes: AttributeTypes | OldAttributeTypes): AttributeTypes {
	if (!attributeTypes || Array.isArray(attributeTypes.nodes) || Array.isArray(attributeTypes.edges)) {
		return {
			nodes: {},
			edges: {}
		}
	}

	return {
		nodes: attributeTypes.nodes ?? {},
		edges: attributeTypes.edges ?? {}
	}
}

function potentiallyUpdateBlacklistTypes(blacklist): BlacklistItem[] {
	for (const entry of blacklist) {
		if (entry.type === ExportBlacklistType.hide) {
			entry.type = BlacklistType.flatten
		}
	}
	return blacklist
}
