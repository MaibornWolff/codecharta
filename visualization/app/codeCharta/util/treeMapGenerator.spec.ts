import { NodeMetricData, State, CodeMapNode, Node } from "../codeCharta.model"
import { TreeMapGenerator } from "./treeMapGenerator"
import { METRIC_DATA, TEST_FILE_WITH_PATHS, VALID_NODE_WITH_PATH, VALID_EDGES, STATE } from "./dataMocks"
import { klona } from 'klona';
import { NodeDecorator } from "./nodeDecorator"
import { fileWithFixedFolders } from "../ressources/fixed-folders/fixed-folders-example"
import { getCCFile } from "./fileHelper"

describe("treeMapGenerator", () => {
	let map: CodeMapNode
	let state: State
	let metricData: NodeMetricData[]
	let codeMapNode: CodeMapNode
	let isDeltaState

	beforeEach(() => {
		restartSystem()
	})

	function restartSystem() {
		map = klona(TEST_FILE_WITH_PATHS.map)
		NodeDecorator.decorateMapWithPathAttribute(getCCFile("someFile", fileWithFixedFolders))
		state = klona(STATE)
		codeMapNode = klona(VALID_NODE_WITH_PATH)
		metricData = klona(METRIC_DATA)
		isDeltaState = false
		state.dynamicSettings.focusedNodePath = ""
	}

	describe("create Treemap nodes", () => {
		it("only root node", () => {
			map.children = []

			const nodes: Node[] = TreeMapGenerator.createTreemapNodes(map, state, metricData, isDeltaState)

			expect(nodes).toMatchSnapshot()
		})

		it("root node with two direct children", () => {
			map.children[1].children = []

			const nodes: Node[] = TreeMapGenerator.createTreemapNodes(map, state, metricData, isDeltaState)

			expect(nodes).toMatchSnapshot()
		})

		it("root node with two direct children and some grand children", () => {
			const nodes: Node[] = TreeMapGenerator.createTreemapNodes(map, state, metricData, isDeltaState)

			expect(nodes).toMatchSnapshot()
		})

		it("should build the tree map with valid coordinates using the fixed folder structure", () => {
			const nodes = TreeMapGenerator.createTreemapNodes(fileWithFixedFolders.nodes[0], state, metricData, isDeltaState)

			expect(nodes).toMatchSnapshot()
		})
	})

	describe("CodeMap value calculation", () => {
		it("if a node was deleted from previous file it should still be visible and have positive width/length", () => {
			// given map with one node deleted in comparison to previous file
			map.attributes = { myArea: 22, myHeight: 12 }
			map.deltas = {}
			map.children[0].attributes = { myArea: 44, myHeight: 63 }
			map.children[0].deltas = { myArea: 20, myHeight: 0 }
			map.children[1].attributes = { myArea: 0, myHeight: 0 }
			map.children[1].deltas = { myArea: -40, myHeight: -80 }

			state.dynamicSettings.areaMetric = "myArea"
			state.dynamicSettings.heightMetric = "myHeight"
			state.treeMap.mapSize = 1000
			metricData = [
				{ name: "myArea", maxValue: 42 },
				{ name: "myHeight", maxValue: 99 }
			]

			const nodes: Node[] = TreeMapGenerator.createTreemapNodes(map, state, metricData, isDeltaState)

			expect(nodes[2].name).toBe("Parent Leaf")
			expect(nodes[2].width).toBeGreaterThan(0)
			expect(nodes[2].length).toBeGreaterThan(0)
		})

		it("attribute exists, no children", () => {
			map.children = []
			map.attributes = { a: 100 }

			const nodes: Node[] = TreeMapGenerator.createTreemapNodes(map, state, metricData, isDeltaState)

			expect(nodes[0].attributes["a"]).toBe(100)
		})

		it("attribute do not exists, no children", () => {
			map.children = []

			const nodes: Node[] = TreeMapGenerator.createTreemapNodes(map, state, metricData, isDeltaState)

			expect(nodes[0].attributes["b"]).toBe(undefined)
		})

		it("attribute do not exists, multiple children with non existant attributes", () => {
			state.dynamicSettings.heightMetric = "b"
			state.dynamicSettings.areaMetric = "b"
			metricData = [
				{ name: "a", maxValue: 42 },
				{ name: "b", maxValue: 99 }
			]

			const nodes: Node[] = TreeMapGenerator.createTreemapNodes(map, state, metricData, isDeltaState)

			expect(nodes[0].attributes["b"]).toBe(undefined)
		})

		it("area should be zero if metric does not exist", () => {
			state.dynamicSettings.areaMetric = "unknown"
			state.dynamicSettings.heightMetric = "unknown"
			state.fileSettings.edges = VALID_EDGES
			metricData = [{ name: "unknown", maxValue: 100 }]

			const nodes: Node[] = TreeMapGenerator.createTreemapNodes(map, state, metricData, isDeltaState)

			expect(nodes[2].width * nodes[2].length).toEqual(0)
		})
	})

	describe("calculateAreaValue", () => {
		it("should return 0 if node has children, not blacklisted and not only visible in comparison map", () => {
			const actual = TreeMapGenerator["calculateAreaValue"](codeMapNode, state)

			expect(actual).toBe(0)
		})
	})
})
