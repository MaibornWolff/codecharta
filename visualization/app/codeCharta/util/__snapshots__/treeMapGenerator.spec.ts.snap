// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`treeMapGenerator create Treemap nodes only root node 1`] = `
Array [
  Object {
    "attributes": Object {},
    "color": "#666666",
    "deltas": undefined,
    "depth": 0,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 500,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": 48,
      "y": 500,
      "z": -101,
    },
    "isLeaf": true,
    "length": 596,
    "link": undefined,
    "markingColor": null,
    "name": "root",
    "outgoingEdgePoint": Vector3 {
      "x": 48,
      "y": 500,
      "z": 197,
    },
    "path": "/root",
    "visible": true,
    "width": 596,
    "x0": 0,
    "y0": 0,
    "z0": 0,
  },
]
`;

exports[`treeMapGenerator create Treemap nodes root node with two direct children 1`] = `
Array [
  Object {
    "attributes": Object {},
    "color": "#666666",
    "deltas": undefined,
    "depth": 0,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 2,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": 83.13843876330611,
      "y": 2,
      "z": -83.43078061834694,
    },
    "isLeaf": false,
    "length": 666.2768775266122,
    "link": undefined,
    "markingColor": null,
    "name": "root",
    "outgoingEdgePoint": Vector3 {
      "x": 83.13843876330611,
      "y": 2,
      "z": 249.70765814495917,
    },
    "path": "/root",
    "visible": true,
    "width": 666.2768775266122,
    "x0": 0,
    "y0": 0,
    "z0": 0,
  },
  Object {
    "attributes": Object {
      "functions": 10,
      "mcc": 1,
      "rloc": 100,
    },
    "color": "#DDDDDD",
    "deltas": undefined,
    "depth": 1,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 2,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": 83.13843876330606,
      "y": 4,
      "z": -73.83078061834698,
    },
    "isLeaf": true,
    "length": 627.8768775266121,
    "link": "http://www.google.de",
    "markingColor": null,
    "name": "big leaf",
    "outgoingEdgePoint": Vector3 {
      "x": 83.13843876330606,
      "y": 4,
      "z": 240.1076581449591,
    },
    "path": "/root/big leaf",
    "visible": true,
    "width": 627.8768775266121,
    "x0": 19.200000000000003,
    "y0": 19.200000000000003,
    "z0": 2,
  },
  Object {
    "attributes": Object {},
    "color": "#666666",
    "deltas": undefined,
    "depth": 1,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 500,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": 406.6768775266122,
      "y": 502,
      "z": -73.83078061834698,
    },
    "isLeaf": true,
    "length": 627.8768775266121,
    "link": undefined,
    "markingColor": null,
    "name": "Parent Leaf",
    "outgoingEdgePoint": Vector3 {
      "x": 406.6768775266122,
      "y": 502,
      "z": 240.1076581449591,
    },
    "path": "/root/Parent Leaf",
    "visible": true,
    "width": 0,
    "x0": 656.6768775266122,
    "y0": 19.200000000000003,
    "z0": 2,
  },
]
`;

exports[`treeMapGenerator create Treemap nodes root node with two direct children and some grand children 1`] = `
Array [
  Object {
    "attributes": Object {},
    "color": "#666666",
    "deltas": undefined,
    "depth": 0,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 2,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": 117.57550765359252,
      "y": 2,
      "z": -66.21224617320374,
    },
    "isLeaf": false,
    "length": 735.151015307185,
    "link": undefined,
    "markingColor": null,
    "name": "root",
    "outgoingEdgePoint": Vector3 {
      "x": 117.57550765359252,
      "y": 2,
      "z": 301.3632614803888,
    },
    "path": "/root",
    "visible": true,
    "width": 735.151015307185,
    "x0": 0,
    "y0": 0,
    "z0": 0,
  },
  Object {
    "attributes": Object {
      "functions": 10,
      "mcc": 1,
      "rloc": 100,
    },
    "color": "#DDDDDD",
    "deltas": undefined,
    "depth": 1,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 2,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": -56.61224617320377,
      "y": 4,
      "z": -61.412246173203755,
    },
    "isLeaf": true,
    "length": 338.7755076535925,
    "link": "http://www.google.de",
    "markingColor": null,
    "name": "big leaf",
    "outgoingEdgePoint": Vector3 {
      "x": 291.76326148038873,
      "y": 4,
      "z": -61.412246173203755,
    },
    "path": "/root/big leaf",
    "visible": true,
    "width": 696.751015307185,
    "x0": 19.200000000000003,
    "y0": 19.200000000000003,
    "z0": 2,
  },
  Object {
    "attributes": Object {},
    "color": "#666666",
    "deltas": undefined,
    "depth": 1,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 2,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": -56.61224617320377,
      "y": 4,
      "z": 296.56326148038875,
    },
    "isLeaf": false,
    "length": 338.77550765359246,
    "link": undefined,
    "markingColor": null,
    "name": "Parent Leaf",
    "outgoingEdgePoint": Vector3 {
      "x": 291.76326148038873,
      "y": 4,
      "z": 296.56326148038875,
    },
    "path": "/root/Parent Leaf",
    "visible": true,
    "width": 696.751015307185,
    "x0": 19.200000000000003,
    "y0": 377.17550765359255,
    "z0": 2,
  },
  Object {
    "attributes": Object {
      "functions": 100,
      "mcc": 100,
      "rloc": 30,
    },
    "color": "#820E0E",
    "deltas": undefined,
    "depth": 2,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 2,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": -119.56734770392222,
      "y": 6,
      "z": 221.4693845669907,
    },
    "isLeaf": true,
    "length": 300.37550765359236,
    "link": undefined,
    "markingColor": null,
    "name": "small leaf",
    "outgoingEdgePoint": Vector3 {
      "x": -119.56734770392222,
      "y": 6,
      "z": 371.65713839378685,
    },
    "path": "/root/Parent Leaf/small leaf",
    "visible": true,
    "width": 184.06530459215554,
    "x0": 38.400000000000006,
    "y0": 396.3755076535926,
    "z0": 4,
  },
  Object {
    "attributes": Object {
      "functions": 1000,
      "mcc": 10,
      "rloc": 70,
    },
    "color": "#DDDDDD",
    "deltas": undefined,
    "depth": 2,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 2,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": 105.4367322709129,
      "y": 6,
      "z": 296.56326148038875,
    },
    "isLeaf": true,
    "length": 300.37550765359236,
    "link": undefined,
    "markingColor": null,
    "name": "other small leaf",
    "outgoingEdgePoint": Vector3 {
      "x": 332.97958762842757,
      "y": 6,
      "z": 296.56326148038875,
    },
    "path": "/root/Parent Leaf/other small leaf",
    "visible": true,
    "width": 455.08571071502945,
    "x0": 241.66530459215554,
    "y0": 396.3755076535926,
    "z0": 4,
  },
  Object {
    "attributes": Object {},
    "color": "#666666",
    "deltas": undefined,
    "depth": 2,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 500,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": 456.351015307185,
      "y": 504,
      "z": 221.4693845669907,
    },
    "isLeaf": true,
    "length": 300.37550765359236,
    "link": undefined,
    "markingColor": null,
    "name": "empty folder",
    "outgoingEdgePoint": Vector3 {
      "x": 456.351015307185,
      "y": 504,
      "z": 371.65713839378685,
    },
    "path": "/root/Parent Leaf/empty folder",
    "visible": true,
    "width": 0,
    "x0": 706.351015307185,
    "y0": 396.3755076535926,
    "z0": 4,
  },
]
`;
